hyper/http20/connection.py

```diff
@@ -161,9 +182,9 @@ def __init__(self, host, port=None, secure=None, window_manager=None,
         #
         # Either _read_lock or _write_lock may be acquired whilst holding _lock
         # which should always be acquired before either of the other two.
-        self._lock = threading.RLock()
-        self._write_lock = threading.RLock()
         self._read_lock = threading.RLock()
+        self._write_lock = _OrderedRLock(self._read_lock)
+        self._lock = _OrderedRLock(self._write_lock)
```

Comment:
There's a bigger problem here, which we need to address. That's the fact that this necessarily means that now to take the `_read_lock` you have to take *all* the locks: that means that our *outer* lock, which normally exists to act as the minimum level of contention, is now impossible to hold when someone is holding the write lock. That rather defeats the point of having this level of lock granularity.

I think we need another approach here. =(